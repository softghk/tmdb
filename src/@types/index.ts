export * from './movie'

export interface ResponsePagination<T> {
  page: number
  results: Array<T>
  total_pages: number
  total_results: number
}
