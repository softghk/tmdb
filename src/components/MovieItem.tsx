import {MovieItemType} from '@types'
import Image from 'next/image'
import React from 'react'
import {getImage} from '@utils'
import StarIcon from '@mui/icons-material/Star'
import StarIconOutline from '@mui/icons-material/StarBorderOutlined'
import {Button} from '@mui/material'

export interface MovieItemProps {
  item: MovieItemType
  isFavorite: boolean
  handleFavorite: (id: number) => void
}

const MovieItem = ({item, isFavorite, handleFavorite}: MovieItemProps) => {
  if (item.poster_path) {
    return (
      <a
        href={`${process.env.NEXT_PUBLIC_BASE}movie/${item.id}`}
        target='_blank'
        className={`border-zinc-800 select-none p-2 border md:hover:scale-105 hover:shadow-lg transition-all rounded-lg overflow-hidden ${
          isFavorite ? 'bg-blue-200' : 'bg-white'
        }`}
      >
        <Image
          loader={() => 'Loading ...'}
          src={getImage(item.poster_path)}
          alt={item.title}
          width={342}
          priority
          className='w-full rounded-lg'
          height={400}
        />
        <div className='p-1'>
          <div className='flex justify-between text-xs text-neutral-400'>
            <span>{item.vote_average > 0 && <>IMDB: {item.vote_average}</>}</span>
            <span>{item.release_date?.slice(0, 4)}</span>
          </div>
          <div className='flex justify-between items-start'>
            <h5>{item.title}</h5>
            {/* Favorite Button */}
            <Button
              onClick={(e) => {
                e.preventDefault()
                handleFavorite(item.id)
              }}
              size='small'
              color='warning'
              data-testid='favorite'
            >
              {isFavorite ? <StarIcon color='warning' /> : <StarIconOutline color='warning' />}
            </Button>
          </div>
        </div>
      </a>
    )
  }

  return null
}

export default MovieItem
