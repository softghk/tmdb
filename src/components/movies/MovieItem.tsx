import {MovieItemType} from '@types'
import Image from 'next/image'
import React, { useState } from 'react'
import {getImage} from 'src/utils/getImage'

interface MovieItemProps {
  item: MovieItemType
}

export const MovieItem = ({item}: MovieItemProps) => {
  const [isStared, setIsStared] = useState(false)
  if (item.poster_path) {
    return (
      <a
        href={`${process.env.NEXT_PUBLIC_BASE}movie/${item.id}`}
        target='_blank'
        className='border-zinc-800 border active:scale-95 hover:scale-105 hover:shadow-lg transition-all rounded-lg overflow-hidden'
      >
        <Image
          loader={() => 'Loading ...'}
          src={getImage(item.poster_path)}
          alt={item.title}
          width={300}
          className='w-full'
          height={300}
        />
        <button></button>
        <div className='p-1'>
          <div className='flex justify-between text-xs text-neutral-400'>
            <span>{item.vote_average > 0 && <>IMDB: {item.vote_average}</>}</span>
            <span>{item.release_date?.slice(0, 4)}</span>
          </div>
          <h5>{item.title}</h5>
        </div>
      </a>
    )
  }

  return null
}
