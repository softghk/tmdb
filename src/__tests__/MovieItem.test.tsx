import {fireEvent, render, screen} from '@testing-library/react'
import {MovieItemType} from '@types'
import MovieItem, {MovieItemProps} from '../components/MovieItem'

describe('MovieItem', () => {
  const item: MovieItemType = {
    adult: false,
    backdrop_path: '/backdrop.jpg',
    genre_ids: [12, 14],
    id: 123,
    original_language: 'en',
    original_title: 'Mock Movie',
    overview: 'This is a mock movie.',
    popularity: 7.5,
    poster_path: '/poster.jpg',
    release_date: '2021-10-01',
    title: 'Mock Movie',
    video: false,
    vote_average: 8.2,
    vote_count: 100,
  }

  const props: MovieItemProps = {
    handleFavorite: () => null,
    isFavorite: true,
    item,
  }

  it('renders Movie Item unchanged', () => {
    const {container} = render(<MovieItem {...props} />)
    expect(container).toMatchSnapshot()
  })
  it('renders correctly', async () => {
    const {getByText} = render(<MovieItem {...props} />)
  })
  it('calls handleFavorite function on button click', () => {
    const mockHandleFavorite = jest.fn()
    render(<MovieItem {...props} handleFavorite={mockHandleFavorite} />)
    const favoriteButton = screen.getByRole('button')
    fireEvent.click(favoriteButton)
    expect(mockHandleFavorite).toHaveBeenCalledWith(123)
  })
  it('renders Image component with correct props', () => {
    const {getByAltText} = render(<MovieItem {...props} />)
    const image = getByAltText(item.title) as HTMLImageElement
    expect(image.alt.includes(item.title)).toBeTruthy()
  })
})
