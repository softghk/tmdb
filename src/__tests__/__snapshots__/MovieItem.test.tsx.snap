// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem renders Movie Item unchanged 1`] = `
<div>
  <a
    class="border-zinc-800 select-none p-2 border md:hover:scale-105 hover:shadow-lg transition-all rounded-lg overflow-hidden bg-blue-200"
    href="https://www.themoviedb.org/movie/123"
    target="_blank"
  >
    <img
      alt="Mock Movie"
      class="w-full rounded-lg"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="400"
      src="Loading ..."
      srcset="Loading ... 1x, Loading ... 2x"
      style="color: transparent;"
      width="342"
    />
    <div
      class="p-1"
    >
      <div
        class="flex justify-between text-xs text-neutral-400"
      >
        <span>
          IMDB: 
          8.2
        </span>
        <span>
          2021
        </span>
      </div>
      <div
        class="flex justify-between items-start"
      >
        <h5>
          Mock Movie
        </h5>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textWarning MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textWarning MuiButton-sizeSmall MuiButton-textSizeSmall css-1n582q2-MuiButtonBase-root-MuiButton-root"
          data-testid="favorite"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeMedium css-3adcp6-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </a>
</div>
`;
